### ISSUE TITLE
Class and decorator-based API

### ISSUE THREAD

**Issue by phillipskevin** [canjs/can-element#55]
This is a proposal for a new can-element API designed around classes and decorators.

This is what the [Beginner CRUD Guide](https://canjs.com/doc/guides/crud-beginner.html) would look like using this API.

```js
@customElementsDefine("todos-app")
@stache()
class TodosApp extends CanElement {
  view = `
    <h1>Today’s to-dos</h1>
    {{# if(this.todosPromise.isPending) }}
      Loading todos…
    {{/ if }}
    {{# if(this.todosPromise.isRejected) }}
      <p>Couldn’t load todos; {{ this.todosPromise.reason }}</p>
    {{/ if }}
    {{# if(this.todosPromise.isResolved) }}
      <input placeholder="What needs to be done?" value:bind="this.newName" />
      <button on:click="this.save()" type="button">Add</button>
      <ul>
        {{# for(todo of this.todosPromise.value) }}
          <li>
            <input checked:bind="todo.complete" on:change="todo.save()" type="checkbox" />
            {{# eq(todo, this.selected) }}
              <input focused:from="true" on:blur="this.saveTodo(todo)" value:bind="todo.name" />
            {{ else }}
              <span class="{{# if(todo.complete) }}done{{/ if }}" on:click="this.selected = todo">
                {{ todo.name }}
              </span>
            {{/ eq }}
            <button on:click="todo.destroy()" type="button"></button>
          </li>
        {{/ for }}
      </ul>
    {{/ if }}
  `

  @coerceType(MaybeString)
  newName = null;

  @coerceType(Todo)
  selected = null;

  get todosPromise() {
    return Todo.getList({sort: "name"});
  }

  save() {
    const todo = new Todo({name: this.newName});
    todo.save();
    this.newName = "";
  }

  saveTodo(todo) {
    todo.save();
    this.selected = null;
  }
}
```

**Comment by phillipskevin** [comment]
A few details...

## stache

The `stache` decorator would allow you to pass a property name so you can use something other than `view`. This would look like `@stache("render")`. The value of this property can either be a string or a function so that you can load external template files

```js
import todosView from "./view.stache";

@stache("template")
class TodosApp extends CanElement {
  template = todosView;

  ...
}
```

## viewRoot

You can customize where the view is rendered by setting a property called `viewRoot`. This will default to rendering the view into the element, but you can set it to a shadowRoot if you want to use shadow DOM.

```js
constructor() {
  var shadow = this.attachShadow({mode: 'open'});
  this.viewRoot = shadow;
}

```

**Comment by justinbmeyer** [comment]
Is this syntax supported in any browser?  What would a supported syntax, without babel, look like?

-------


Could `stache` be added like:

```js
class TodosApp extends CanElement {
  @stache()
  view = `...`
```

This way, the property could be known?

---------

What we need to be imported?  I see:

```js
import {customElementsDefine, stache, CanElement, coerceType, MaybeString}
```

I sorta lean towards a `StacheElement` and maybe `type` import, so we could go:

```js
import {customElementsDefine, StacheElement, type} from "can/everything";

@customElementsDefine("todos-app")
class TodosApp extends CanElement {
  view = `...`

  type.coerce(type.MaybeString)
  newName = null;
}
```

I also notice that `Todo` isn't defined.  We should probably mock up what that looks like to make sure stuff feels right together.

**Comment by phillipskevin** [comment]
Supported without babel you mean?

**Comment by justinbmeyer** [comment]
Yes, supported w/o babel.

**Comment by phillipskevin** [comment]
You can do the decorators around the class without babel (althought it's ugly).

```js
var TodosApp = (function() {
  class TodosApp extends CanElement {
    ...
  }

  return customElementsDefine("todos-app")( stache()(TodosApp) );
}());
```

Making this possible is the reason the `stache` decorator was not on the view property itself, although I prefer `@stache() view = "...";` personally.

We could make a `StacheElement` like you mention and just show using the normal `customElements.define( ... )` without a decorator.

The type decorators are the harder part.

**Comment by phillipskevin** [comment]
> I also notice that Todo isn't defined.

I was just using the one from the Guide:

```js
const Todo = realtimeRestModel("/api/todos/{id}").Map;
```

but I can write this up using `can-observe`.

**Comment by phillipskevin** [comment]
We could show it like this with classes and decorators:

```js
import { stacheDecorator, customElement, CanElement } from "can";

@customElement("todos-app")
@stacheDecorator()
class TodosApp extends CanElement {
  view = ` ... `;
}
```

And then have an "ES5 Setup" (or something like that) section that shows using:

```js
import { StacheElement } from "can";

class TodosApp extends StacheElement {
  view() {
    return ` ... `;
  }
}
customElements.define("todos-app", TodosApp);
```

**Comment by matthewp** [comment]
`@customElementsDefine("todos-app")` I don't think this one really gives you much.

```js
@customElementsDefine("todos-app")
class TodosApp extends StacheElement {}
```

vs.

```js
class TodosApp extends StacheElement {}

customElements.define("todos-app", TodosApp);
```

I'd rather just do the latter.

**Comment by matthewp** [comment]
There's some confusion here, decorators are exactly the same as ES6 class mixins. So:

```js
@stache
class TodosApp extends HTMLElement {}
```

And 

```js
class TodosApp extends stache(HTMLElement) {}
```

Are the same.

**Comment by christopherjbaker** [comment]
@matthewp Unless there's some magic going on behind the scenes, the second example above suggests to me that the stache method will be called with HTMLElement and is expected to return a class that can be extended. The first option, stache will be called with a class which extends HTMLElement and is expected to modify it or return a new class.

If these are in fact the same, will the class mixins signature change to match the new decorator signature?

**Comment by matthewp** [comment]
@christopherjbaker If I'm understanding you correctly, you are saying that the decorator version receives an already extended `TodosApp` class, whereas the first does not. If so then you are correct that they are not exactly the same.

My reason for commenting was [this comment](https://github.com/canjs/can-element/issues/55#issuecomment-488440827) suggested that class mixin approach was ugly and I don't agree with that all, it's just functions returning functions.

----

To summarize, @phillipskevin is correct that this is the ES6 equivalent to how decorators work, but you don't need to do it that way and wouldn't.

**Comment by christopherjbaker** [comment]
The decorator does receive an already extended TodosApp, but I am unfamiliar with the mixin. The syntax *suggests* to me that it works how I mentioned, but I do not know.

**Comment by mjstahl** [comment]
TL;DR If we are still attempting to brand as "the quickest way to build CRUD apps", optimize for that case by removing as many traces of the framework as possible. The closer to vanilla JS the better and that means less choice and more picking defaults. The closer to vanilla JS the less time needed to think about how the framework would do things which means more time spent thinking about the problem domain.

With that said, I like 

```js
import { StacheElement } from "can";

export default class TodosApp extends StacheElement {
  // by default, the render function name
  render() {
    // by default, expect a stache string to be returned from this function
    return ` ... `;
  }
}
// just DOM
customElements.define("todos-app", TodosApp);
```

In regards to default names, I prefer verbs for functions and nouns for properties (which is why I named the above function `render`). However, in this example, the `StacheElement` base class could have some function that looks for a `view` property as opposed to having a function. I am not sure which way I feel better about, but the property route is less typing (for what that is worth).

I am still thinking through how/what I feel about properties, type coercion, `value` (all the other DefineMap stuff). So I have just a couple of opinions of type coercion. 

I didn't know what 'MaybeString' was and had to look it up. If we other type objects like that through the framework, they should definitely be united so that they can be included by a simple import. So in the case, I think @justinbmeyer is correct when doing `types.MaybeString`, I am not sure how I feel about the `coerce` function being a part of that package as well.

I believe the `coerce` function is important because it relieves the programmer for that one extra decision of having to use `type` or `Type`.

**Comment by justinbmeyer** [comment]
How do people feel about throwing away `class` here?

```js
import {stacheElement} from "can/everything";

stacheElement.define("my-tag",{
  view: ``,
  properties: {
    foo: String
  },
  methods: {

  }
})
```

How much is `class` worth?  Imo, it's mostly worth it b/c of `observe.Object` and making models.

Another idea:

```js
const MyElement = element("my-element",function( {first, last} ){

	var fullName = value.returnedBy( () => first.value + " " + last.value );

	return stache`<input value="${ first }" 
		onChange=${ (ev)=> first.value = ev.currentTarget.value }/> ${fullName}`
})
```

**Comment by mjstahl** [comment]
@justinbmeyer `class` is worth it because it gives us the semantics, through inheritance, to abstract a lot of the framework away from the User.

**Comment by matthewp** [comment]
@justinbmeyer Can't as easily mixin different behaviors that way, would have to come up with your own mixin pattern.

**Comment by christopherjbaker** [comment]
I think getting away from classes would be a very bad idea. These new syntaxes were introduced so that devs would not need their own mixin patterns, but could use common patterns instead. Even if classes are not ideal, I think we should use them just to make it easier on other devs.

**Comment by phillipskevin** [comment]
Here's a summary of the APIs I'm thinking after the discussion above. Let me know what you think.

`StacheElement`

This is an extended HTMLElement with:
- its prototype [proxied with can-observe](https://github.com/canjs/can-observe/pull/80) so changes to the element's properties will dispatch events and rerender
- [lifecycle methods](https://github.com/canjs/can-element/issues/54) that make testing easier and prevent events from being dispatched during initialization
- stache as the default renderer / templating engine

`types`

A package of [data-types](https://canjs.com/doc/can-data-types.html) that can be used as decorators on class fields (also possible to use in non-babel environments - discussed below).

**Here's an example of what this would look like put together:**

```js
import { StacheElement, types } from "can";

class MyEl extends StacheElement {
  render() {
    return `<h1>{{fullName}}</h1>`;
  }

  @types.MaybeString.coerce
  firstName = "";

  @types.MaybeString.coerce
  lastName = "";

  get fullName() {
    return `${this.firstName} ${this.lastName}`;
  }
}
customElement.define("my-el", MyEl);
```

## Non-babel use-case

The only difference in environments without public fields and decorators, is you would need to call the `type` decorators as functions as discussed [in this PR](https://github.com/canjs/can-observe/pull/69).

```js
import { StacheElement, types } from "can";

class MyEl extends StacheElement {
... 
}

types.MaybeString.coerce(MyEl, "firstName");
types.MaybeString.coerce(MyEl, "lastName");
```

**Comment by phillipskevin** [comment]
I've gone back and forth on using a `render` function vs using a `view` property.

@christopherjbaker pointed out that specifying stache templates like I've shown above might be confusing. Since it is called `render` and you will almost always return a template literal, people may think they can do something like this:

```js
class MyEl extends StacheElement {
  render() {
    return `<h1>${this.fullName}</h1>`;
  }
}
```

This will actually work for the initial render, but this function won't be called again if `fullName` changes.

This has convinced me that we should use a `view` property to make it clear that it is not dynamic and make it impossible to use `this` inside of the template literal.

For all of the people that use transpilers, we can show this using a public field:

```js
class MyEl extends StacheElement {
  view = `...`;
}
```

For environments that don't support public fields (only supported in Chrome or with Babel currently), we could show setting the view property in the constructor.

```js
class MyEl extends StacheElement {
  constructor() {
    super();
    this.view = `...`;
  }
}
```

We could also allow people to use a getter. This makes it more clear (than a `render` function) that this is a property, but doesn't 100% resolve the confusion since other getters in `StacheElement`s can access properties on `this` and will re-execute when those properties change.

```js
class MyEl extends StacheElement {
  get view() {
    return `...`;
  }
}
```

**Comment by matthewp** [comment]
@phillipskevin In can-component the view is static. Why wouldn't it be so here? I think we need to hold to this assumption, that way StacheElement doesn't have to call `stache(string)` on every element instance.

**Comment by phillipskevin** [comment]
Yeah, that's a good point. So you think people should do this?

```js
class MyEl extends StacheElement { }
MyEl.view = ` ... `;
```

and also support

```js
class MyEl extends StacheElement {
  static get view() {
    return `...`;
  }
}
```

**Edit:** this is what he meant

```js
class MyEl extends StacheElement {
  static view = `...`;
}
```

`static get view() { }` would be fine also

**Comment by BigAB** [comment]
I think Mark is right in the 'The closer it is to just normal JS conventions the better' or however he said it, that combined with a static view property looks pretty good to me:

```
import { StacheElement } from "can";

class TodosApp extends StacheElement {
  static view = ` ... `

  @coerceType(MaybeString)
   name = null;
  
  get todosPromise() {
    return Todo.getList({sort: "name"});
  }
}
// just DOM
customElements.define("todos-app", TodosApp);
```

But with the latest [new-new decorators standard](https://github.com/tc39/proposal-decorators), I don't think there is a nice way to use the decorator as a function. So I don't think [this PR](https://github.com/canjs/can-observe/pull/69) provides a good non-decorator solution anymore.

Without decorators, is there a nice looking way to enhance properties with type coercion/checking?

**Comment by justinbmeyer** [comment]
@mjstahl 

> @justinbmeyer class is worth it because it gives us the semantics, through inheritance, to abstract a lot of the framework away from the User.

I'm not sure what you mean. Can you elaborate?  If you just mean inheritance, I don't this `extends` is adding that much value.  VueJS does not use class.  

> I think getting away from classes would be a very bad idea. These new syntaxes were introduced so that devs would not need their own mixin patterns, but could use common patterns instead. Even if classes are not ideal, I think we should use them just to make it easier on other devs.

By "these new syntaxes", do you mean decorators?

What are we exactly making easy? VueJS does not use class. At some point using class makes things harder on devs compared to using object literals.

To be clear, I want classes and decorators to work, but there's something I want much more - __to be able to use CanJS `reasonably` without having a transpiler__.  If we are going to accept a transpiler, we should then be exploring crazy ideas like SvelteJS.  

We should probably have a larger discussion around goals and use that to inform design.  Here's what I'm thinking for CanJS 6 in terms of priorities:

1. Make CanJS easy for users to learn and build apps
2. Build to web components natively
3. Allow CanJS to be used `reasonably` without a transpiler.
4. Allow CanJS to be lighter weight.
5. Have CanJS appear modern

I don't think `class` is needed to make a framework that is easy for users to learn and build apps. As I noted above, VueJS does not use class.  It uses object literals. React more or less uses functions. `class` can obviously help make the following easier to understand:

- inheritence
- `super`
- constructor 

But CanJS isn't really about inheritance. It's about describing the logic of various properties. That is the majority of what users do and we should optimize around making that as straightforward as possible.

All this is to say, I don't think we should get hung up on `class`, especially as `decorators` has been in constant flux and if we want to have CanJS work without a transpiler.

**Comment by matthewp** [comment]
@justinbmeyer 

Note that [Vue 3](https://medium.com/the-vue-point/plans-for-the-next-iteration-of-vue-js-777ffea6fabf) will be based on classes.

I like your list of priorities for the most part. Most of those are covered by can-component, however. What is the purpose of can-element if it is only a slightly different API vs. can-component.

Also, missing from your list is making CanJS be a set of tool for building web components (which we talked about being important to you last week). I think that's the big one where classes make a lot more sense. can-component, or any other can-construct based solution doesn't have a model for mixins.

I think there's room here for a middle ground where classes can be used without transpilers or without depending on the still in-flux decorator idea. Classes are a bit more restrictive than object literals, but you can use object literals inside of class bodies. You just can't put them on the prototype. That's ok though, there's no reason for property definitions to be on the prototype since they don't inherit that way anyways.

Looking through this thread it seems that there are just a few places where decorators help:

* Types that can be coerced or be strict.
* Resolver types of functions.

Can we come up with a way to do these same things with classes that all browsers already support? Here's a very rough way to do that, which could be refined.

```js
class Panel extends StacheElement {
  static view = ``
  static properties = {
    title: String
  }
  static resolvers = {
    count({ resolve, listenTo }) {
      // do things
    }
  }
   methodOne() {
   }
}
```

Resolvers are a bit special to CanJS but other frameworks have ways to coerce property types and do it similar to above. React has [propTypes for example](https://reactjs.org/docs/typechecking-with-proptypes.html).

It would also be easy to build an object-literal API on top of this if you really wanted that, but without giving up the ability to mixin different behaviors or to use different view libraries.

**Comment by justinbmeyer** [comment]
> Note that Vue 3 will be based on classes.

I was pretty gung ho on making CanJS use classes too. But the churn on decorators make this far less appealing.  But perhaps I'm misdirecting my ire at classes when I should be directing it at decorators.

> missing from your list is making CanJS be a set of tool for building web components

That was somewhat encapsulated in #2, though not stated explicitly.

> doesn't have a model for mixins

I'm not sure `class` currently has a model for mixins. There's ideas, but none of them are in a browser.

-------


I like the `static` idea.  These are expressions, so it should be possible to do all the type stuff we want.

**Comment by matthewp** [comment]
How close is the stache expression parser to being able to pull something like this off?

```js
const NameElement = stacheElement(`
  {{#reactive}}
    first = 'Matthew'
    last = 'Phillips'
    fullName = first + ' ' + last
  {{/reactive}}

  <h1>{{fullName}}</h1>

  <input type="text" value:bind="first">
  <input type="text" value:bind="last">
`);

document.body.append(new NameElement());
```

**Comment by justinbmeyer** [comment]
@matthewp Semi close, but I'd hack Babel to do something like that.  Then you'd have something like svelte.

**Comment by matthewp** [comment]
Another thing that occurred to me, is there any reason why things like resolvers and async getters can't be defined as normal functions? For example, instead of (from above)

```js
class Foo extends StacheElement {
  static resolvers = {
    count({ resolve, listenTo }) {
      // do things
    }
  }
}
```

Why not:

```js
class Foo extends StacheElement {
  count() {
    let { resolve, listenTo } = this.resolver();
  }
}
```

I sort of like this better. Trying to make all of the APIs defined through metadata is sort of fighting against the constraints of classes.  I'd still do type definitions like that though.

In other words, if declarative APIs are a weakness of classes why not use imperative ones instead.

**Comment by justinbmeyer** [comment]
@matthewp how would that work?  Would count be accessed as a function?  

```
var foo = new Foo()
foo.count()
```

Something like this would work:

```js
class Foo extends StacheElement {
  get  count() {
    return resolve(this, "count",(resolve, listenTo )=>{} )
  }
}
```

It would essentially replace the `getter` (like can-define-lazy) with a getter connected to the observable.  If we had something "stateful" ... and changed the prototype getter / setters to "announce" which property was being read, we could shorten it to:

```js
class Foo extends StacheElement {
  get  count() {
    return resolve(this, (resolve, listenTo )=>{} )
  }
}
```

Actually, if we are announcing the property, we might as well announce the `this` too:

```js
class Foo extends StacheElement {
  get  count() {
    return resolve( (resolve, listenTo )=>{} )
  }
}
```

**Comment by matthewp** [comment]
Do we need the callback function? Can resolve() not just return the listenTo/resolve? I know why maybe not, would be nicer if it could though

**Comment by justinbmeyer** [comment]
Might be possible. I like this. Types are the big problem as they need to be in place before an instance is created. Lazy won’t work.

**Comment by BigAB** [comment]
How about a defineProperties symbol, exported from can, that is basically similar to the define property of can.Map, that could be defined statically

```
import { StacheElement, defineProperties as properties } from 'can';

class TodosApp extends StacheElement {

  static [properties]  = {
    newName: { type: String },
    selected: { type: Todo },
  }

  static view = `
    <h1>Today’s to-dos</h1>
    {{# if(this.todosPromise.isPending) }}
      ...
    {{/ if }}
  `;

  save() {
    const todo = new Todo({name: this.newName});
    todo.save();
    this.newName = "";
  }

  saveTodo(todo) {
    todo.save();
    this.selected = null;
  }
}

// just DOM
customElements.define('todos-app', TodosApp);
```

Would stay familiar to developers, easy to conceptualize what's "going on" in the super classes, and we could still offer the decorator methods to anyone who wanted to use them.

**Comment by phillipskevin** [comment]
Issue moved to [canjs/can-stache-define-element #3](https://github.com/canjs/can-stache-define-element/issues/3) via [**ZenHub**](https://www.zenhub.com/)
